### ISSUE TITLE
Maximising FFmpegFrameGrabber and FFmpegFrameRecorder performance 

### ISSUE THREAD

**Issue by b005t3r** [bytedeco/javacv#1163]
I'd like to know if there's anything I can do to make these run faster. With 4K videos, I don't even get real-time performance which would allow to play back the video normally (I get about 26 FPS for a 30 FPS video), when I throw in encoding, I get something between 6 FPS and 12 FPS, depending on params used.

This is the code I use for testing:

```java
FFmpegFrameGrabber grabber = new FFmpegFrameGrabber("in.mp4");
grabber.start();

FrameRecorder recorder = new FFmpegFrameRecorder("out.mp4", grabber.getImageWidth(), grabber.getImageHeight());
recorder.setFormat(grabber.getFormat());
recorder.setPixelFormat(AV_PIX_FMT_YUV420P);
recorder.setFrameRate(grabber.getFrameRate());
recorder.setVideoBitrate(grabber.getVideoBitrate());
recorder.setVideoCodec(grabber.getVideoCodec());
recorder.setVideoOption("preset", "ultrafast");
recorder.setVideoCodecName("libx264");
recorder.setVideoCodec(AV_CODEC_ID_H264);
recorder.start();

int frameCount = 0;
long startTime = System.currentTimeMillis();

Frame frame;
while ((frame = grabber.grabFrame(false, true, true, false)) != null) {
    recorder.record(frame);
    ++frameCount;
}

float secs = (System.currentTimeMillis() - startTime) / 1000.0f;
System.out.println("processing " + frameCount + " frames took " + (Math.round(secs * 1000) / 1000.0f) + " (" + (Math.round(frameCount / secs * 1000) / 1000.0f) + " fps)");
```

Is there anything I can do to make things faster? Docs for this are next to non-existent, that's why I'm asking this here.

I'm testing this on MacBook Pro 13" (MacBookPro14,1, Intel Core i7 2,5 GHz).

**Comment by saudet** [comment]
Try to use some hardware accelerator that might be available in your machine and that FFmpeg supports.

**Comment by saudet** [comment]
There's plenty of documentation available for FFmpeg: https://ffmpeg.org/documentation.html
If you don't have time to look it up, you can't expect someone else to do it for you!

**Comment by b005t3r** [comment]
Believe me, man, I've been looking for any tips for the last 8 hours and didn't really find much.

**Comment by b005t3r** [comment]
Also, I checked and it looks like hw acceleration should actually be automatically selected, if available.

**Comment by saudet** [comment]
FFmpeg doesn't automatically use hardware acceleration, no. It looks like FFmpeg supports something called VideoToolbox on Mac: https://trac.ffmpeg.org/wiki/HWAccelIntro

It doesn't however tell us much more than that, but it looks the name of the codec is "h264_videotoolbox":
https://trac.ffmpeg.org/ticket/5357
So try to call `setVideoCodecName("h264_videotoolbox")` before start! If you need more information about that codec, you should probably try it ask it to the author of that extension via a ticket over there or by email.

**Comment by b005t3r** [comment]
Thanks for getting back to me :)

I gave it a go. There is a "h264_videotoolbox" codec available, but only as an encoder. I checked it with JavaCV and when used with ```setVideoCodecName()``` in a recorder, it does speed things up considerably, so that's cool. However there's no such codec for decoding (actually, grabber does a pretty bad job reporting that a given codec hasn't been found - it doesn't report that at all, just falls back to using the first one that does work).

Now, for the command line version of ffmpeg they say you should use "-hwaccel videotoolbox" to enable it, if possible (supported format, etc.), but I have no idea how (and if) this translates to you API.

Also I did a small comparison with the command line tool - the same video, software decoder, videotoolbox encoder, the same video for both and JavaCV version is about 2x slower (it looks like only decoding is much slower actually).

**Comment by saudet** [comment]
As shown in issue #1154 we can call `setOption("hwaccel", "videotoolbox")` to set that option.

**Comment by b005t3r** [comment]
Thanks! It didn't change anything, looks like only software decoding works on Mac, but good to know how to change it.

Now, why is it so slow when compared to the command line tool? It's basically 2x-3x slower. I understand that native calls from Java come at a cost, but this seems a bit too much. I even went through grabber's and recorder's code, but didn't find anything that would slow things down (compared to ffmpeg C examples, everything is basically the same).

Is it maybe because of non-release binaries used in the project (libavcodec, etc.)?

**Comment by saudet** [comment]
Why not just use the ffmpeg program itself if you're satisfied with its results?

**Comment by b005t3r** [comment]
Because I'm not doing a simple transcoding app, obviously :) I'm using the command line tool for comparison, that's all.

Is there a reason you can think of why it's so much faster?

**Comment by saudet** [comment]
Compare the logs that it prints on the console.

**Comment by b005t3r** [comment]
I did, but there's nothing there.

**Comment by saudet** [comment]
Make sure to call FFmpegLogCallback.set().

**Comment by b005t3r** [comment]
By "there's nothing there" I meant that there's nothing interesting there :)

Here's the output from my app:

`objc[1683]: Class JavaLaunchHelper is implemented in both /Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/bin/java (0x10e2864c0) and /Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/libinstrument.dylib (0x10e3064e0). One of the two will be used. Which one is undefined.
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from '/Users/booster/Downloads/20171110_151932.mp4':
  Metadata:
    major_brand     : mp42
    minor_version   : 0
    compatible_brands: isommp42
    creation_time   : 2017-11-10T14:19:45.000000Z
    location        : +26.2233-080.1752/
    location-eng    : +26.2233-080.1752/
    com.android.version: 7.0
  Duration: 00:00:12.92, start: 0.000000, bitrate: 17287 kb/s
    Stream #0:0(eng): Video: h264 (High) (avc1 / 0x31637661), yuv420p(tv, bt709), 1920x1080, 17027 kb/s, SAR 1:1 DAR 16:9, 30.02 fps, 30 tbr, 90k tbn, 180k tbc (default)
    Metadata:
      creation_time   : 2017-11-10T14:19:45.000000Z
      handler_name    : VideoHandle
    Stream #0:1(eng): Audio: aac (LC) (mp4a / 0x6134706D), 48000 Hz, stereo, fltp, 256 kb/s (default)
    Metadata:
      creation_time   : 2017-11-10T14:19:45.000000Z
      handler_name    : SoundHandle
[h264_videotoolbox @ 0x7fb17ff22c00] Color range not set for yuv420p. Using MPEG range.
Output #0, mp4, to '/Users/booster/Downloads/out.mp4':
  Metadata:
    encoder         : Lavf58.20.100
    Stream #0:0: Video: h264 (Constrained Baseline) (avc1 / 0x31637661), yuv420p, 1920x1080, q=2-31, 17027 kb/s, 30.02 fps, 739721.00 tbn, 30.02 tbc`

And here's the output from the command line tool:

`MacBook-Pro:Downloads booster$ ffmpeg -hide_banner -hwaccel videotoolbox -i /Users/booster/Downloads/20171110_151932.mp4 -codec:v h264_videotoolbox /Users/booster/Downloads/out2.mp4
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from '/Users/booster/Downloads/20171110_151932.mp4':
  Metadata:
    major_brand     : mp42
    minor_version   : 0
    compatible_brands: isommp42
    creation_time   : 2017-11-10T14:19:45.000000Z
    location        : +26.2233-080.1752/
    location-eng    : +26.2233-080.1752/
    com.android.version: 7.0
  Duration: 00:00:12.92, start: 0.000000, bitrate: 17287 kb/s
    Stream #0:0(eng): Video: h264 (High) (avc1 / 0x31637661), yuv420p(tv, bt709), 1920x1080, 17022 kb/s, SAR 1:1 DAR 16:9, 30.02 fps, 30 tbr, 90k tbn, 180k tbc (default)
    Metadata:
      creation_time   : 2017-11-10T14:19:45.000000Z
      handler_name    : VideoHandle
    Stream #0:1(eng): Audio: aac (LC) (mp4a / 0x6134706D), 48000 Hz, stereo, fltp, 256 kb/s (default)
    Metadata:
      creation_time   : 2017-11-10T14:19:45.000000Z
      handler_name    : SoundHandle
File '/Users/booster/Downloads/out2.mp4' already exists. Overwrite ? [y/N] y
Stream mapping:
  Stream #0:0 -> #0:0 (h264 (native) -> h264 (h264_videotoolbox))
  Stream #0:1 -> #0:1 (aac (native) -> aac (native))
Press [q] to stop, [?] for help
[h264_videotoolbox @ 0x7faf31005200] Color range not set for nv12. Using MPEG range.
Output #0, mp4, to '/Users/booster/Downloads/out2.mp4':
  Metadata:
    major_brand     : mp42
    minor_version   : 0
    compatible_brands: isommp42
    com.android.version: 7.0
    location        : +26.2233-080.1752/
    location-eng    : +26.2233-080.1752/
    encoder         : Lavf58.12.100
    Stream #0:0(eng): Video: h264 (h264_videotoolbox) (avc1 / 0x31637661), nv12, 1920x1072 [SAR 1:1 DAR 120:67], q=2-31, 200 kb/s, 30 fps, 15360 tbn, 30 tbc (default)
    Metadata:
      creation_time   : 2017-11-10T14:19:45.000000Z
      handler_name    : VideoHandle
      encoder         : Lavc58.18.100 h264_videotoolbox
    Stream #0:1(eng): Audio: aac (LC) (mp4a / 0x6134706D), 48000 Hz, stereo, fltp, 128 kb/s (default)
    Metadata:
      creation_time   : 2017-11-10T14:19:45.000000Z
      handler_name    : SoundHandle
      encoder         : Lavc58.18.100 aac
frame=  388 fps=192 q=-0.0 Lsize=    2856kB time=00:00:12.90 bitrate=1813.8kbits/s speed=6.37x    
video:2641kB audio:203kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: 0.434555%
[aac @ 0x7faf31006400] Qavg: 218.288` 

Basically, transcoding works 3 times faster in the command line app that decoding on its own using JavaCV, which is crazy. I don't think it's because JNI's overhead, you don't make that many JNI calls per each ```grabFrame()```.

Is this maybe because of compilation params you used when build ffmpeg? I tried going through your scripts yesterday, but couldn't find it what exactly has been used. Could you post that here?

**Comment by saudet** [comment]
It looks like you're not reencoding the audio when using the command line program. We would need to do the same with JavaCV to be able to compare their performance, see issue #1125.

You can check the ffmpeg program bundled with the presets for the compilation parameters:
https://github.com/bytedeco/javacpp-presets/blob/master/ffmpeg/src/main/java/org/bytedeco/ffmpeg/ffmpeg.java

**Comment by b005t3r** [comment]
I don't really think throwing in audio would make things faster :) Decoding on its own (simply calling `grabFrame()` in a loop) is 3x slower than video + audio transcoding (and yes, I tried it with audio as well, nothing changed).

About the bundled ffmpeg, I tried copy pasting your example, but it didn't work:
```
Error:(7, 56) java: package org.bytedeco.ffmpeg does not exist
```

Anything else I need to add as a dependency?

**Comment by saudet** [comment]
It's a new feature not yet released, you'll need to use the snapshots: http://bytedeco.org/builds/

**Comment by b005t3r** [comment]
OK, how do I use this with gradle?

I currently have this as a dependency:

```
compile group: 'org.bytedeco', name: 'javacv-platform', version: '1.4.4'
```

**Comment by saudet** [comment]
The instructions for Gradle are also on the page:
> The `pom.xml` file above can also be used to download snapshots for others tools such as Gradle or sbt, which are incapable of downloading the snapshot artifacts by themselves. Simply save a dummy `pom.xml` file outside of any project context, run `mvn -U compile` on it, and add the necessary settings for Gradle or sbt to use artifacts from your local Maven repository. In the case of Gradle, [we specify that with `repositories { mavenLocal() }` along with the `dependencies { }` block](https://docs.gradle.org/current/userguide/repository_types.html#example_adding_the_local_maven_cache_as_a_repository), in `app/build.gradle` for an Android project, while for sbt, [we can put `resolvers += Resolver.mavenLocal` next to the `libraryDependencies`](https://www.scala-sbt.org/1.x/docs/Library-Dependencies.html#Resolvers).

**Comment by b005t3r** [comment]
I've read it and yet I don't understand a single thing :)

I've never used Maven in my life and I have no idea what am I supposed to do. I'm guessing I'd need to create a local Maven repo and somehow add it as a dependency? I have no idea how to do it and not having clear instructions doesn't really help that much. 

Anyway, I'll try to build ffmpeg module from javacv-presets on my machine and use ffmpeg tool from there and let you know if it's slower/faster than the official release.

**Comment by b005t3r** [comment]
I checked with the compiled ffmpeg tool - it's as fast as the official release, so about 3x faster than JavaCV 1.4.4.

**Comment by saudet** [comment]
The ffmpeg program probably doesn't convert images to RBG back and forth, and that can be pretty expensive, so if you don't need RGB images, try to call setPixelFormat() as appropriate.

**Comment by b005t3r** [comment]
It's not it, I think I mentioned it before, that I tested grabbing frames without processing and it's just a couple FPS faster (no encoding, just grabbing frames in a loop).

**Comment by saudet** [comment]
Since you're not willing to test it out, I'll assume that's the cause here.

**Comment by b005t3r** [comment]
Dude,I just told you I tested it already. I tested everything you mentioned, changing the pixel format, not doing any pressing on the decoded image, not even encoding it and it still is about 3x slower.

All the things you mentioned and more I already checked.

**Comment by saudet** [comment]
Great! Now that you know how to get around FFmpeg and JavaCV, keep going
like that, let me know when you find the cause, and we'll fix it.

**Comment by b005t3r** [comment]
OK, that was fun. Three days of fun, to be exact. How about reopening this issue and changing it into an actual issue instead of "help wanted"? There's a slight chance someone will actually try to solve it this way, because as it is now, I'm pretty sure it won't ever get noticed.

**Comment by b005t3r** [comment]
BTW, just tested with your `ReadFewFrame.java` example and this example is on par with the command line tool in terms of decoding performance (with the color conversion/`sws_scale()` call). So yeah, there's an issue with grabbers implementation probably.

**Comment by saudet** [comment]
I told you what the issues might be: unneeded image conversion from and to
RGB, needlessly decoding and rencoding audio frames (and there already is
an issue opened for that). That's more than enough to account for the
discrepancies yon mention. You told me you've already tested that, but you
show none of your code, none of your modifications. What do you want me to
say? If you're not motivated enough to solve your own problem, nobody else
will either!

**Comment by b005t3r** [comment]
OK, to make things clear. What do you want me to do to convince you I tested those things? :)

Unneeded color conversion is NOT the issue I'm after. I tested decoding with `sws_scale()` and without it, it doesn't make that much of a difference. I tested the most basic scenario - only decoding video frames, without audio decoding, without any encoding at all, and the results are:
- grabber is about 3x slower than ffmpeg tool (decoding of video frames alone)
- `ReadFewFrame.java` example is on par with `ffmpeg` tool (decoding of video frames alone, with color conversion)

You might open an issue (or I can do it if you want) for optimising grabber's performance, because there clearly is an issue, maybe someone will fix it one day. I'll simply use the code from the other example, because its performance is much better.

**Comment by saudet** [comment]
Where is your modified version of FFmpegFrameGrabber?

**Comment by b005t3r** [comment]
I've been using your implemention, didn't modify it. Why would I change it?

**Comment by b005t3r** [comment]
If you were asking about how my test code looks like, this is what I used to test your grabber (you can uncomment the commented out parts to throw recorded in as well):

```
FFmpegFrameGrabber grabber = new FFmpegFrameGrabber("in.mp4");
grabber.setOption("hwaccel", "videotoolbox");
grabber.start();

/*
FrameRecorder recorder = new FFmpegFrameRecorder("out.mp4", grabber.getImageWidth(), grabber.getImageHeight());
recorder.setFormat(grabber.getFormat());
recorder.setFrameRate(grabber.getFrameRate());
recorder.setVideoBitrate(grabber.getVideoBitrate());
recorder.setVideoCodec(grabber.getVideoCodec());
recorder.setVideoCodecName("h264_videotoolbox");
recorder.start();
*/

int frameCount = 0;
long startTime = System.currentTimeMillis();
Frame frame;
while ((frame = grabber.grabFrame(false, true, true, false)) != null) {
    //recorder.record(frame);
    ++frameCount;
}

float secs = (System.currentTimeMillis() - startTime) / 1000.0f;

System.out.println("processing " + frameCount + " frames took " + (Math.round(secs * 1000) / 1000.0f) + " (" + (Math.round(frameCount / secs * 1000) / 1000.0f) + " fps)");

//recorder.stop();
grabber.stop();
```

And here's `ReadFewFrame.java` I found on your website and just modified a few minor things to measure time and hardcoded video input path:

```
    public static void main(String[] args) throws Exception {
/*
        System.out.println("Read few frame and write to image");

        if (args.length < 1) {
            System.out.println("Missing input video file");
            System.exit(-1);
        }
*/

        int ret = -1, i = 0, v_stream_idx = -1;
        //String vf_path = args[0];
        String vf_path = "/Users/booster/Downloads/20171110_151932.mp4";
        avformat.AVFormatContext fmt_ctx = new avformat.AVFormatContext(null);
        avcodec.AVPacket pkt = new avcodec.AVPacket();

        ret = avformat_open_input(fmt_ctx, vf_path, null, null);
        if (ret < 0) {
            System.out.printf("Open video file %s failed \n", vf_path);
            throw new IllegalStateException();
        }

        // i dont know but without this function, sws_getContext does not work
        if (avformat_find_stream_info(fmt_ctx, (PointerPointer)null) < 0) {
            System.exit(-1);
        }

        av_dump_format(fmt_ctx, 0, vf_path, 0);

        for (i = 0; i < fmt_ctx.nb_streams(); i++) {
            if (fmt_ctx.streams(i).codecpar().codec_type() == AVMEDIA_TYPE_VIDEO) {
                v_stream_idx = i;
                break;
            }
        }
        if (v_stream_idx == -1) {
            System.out.println("Cannot find video stream");
            throw new IllegalStateException();
        } else {
            System.out.printf("Video stream %d with resolution %dx%d\n", v_stream_idx,
                    fmt_ctx.streams(i).codecpar().width(),
                    fmt_ctx.streams(i).codecpar().height());
        }

        avcodec.AVCodecContext codec_ctx = avcodec_alloc_context3(null);
        avcodec_parameters_to_context(codec_ctx, fmt_ctx.streams(v_stream_idx).codecpar());

        avcodec.AVCodec codec = avcodec_find_decoder(codec_ctx.codec_id());
        if (codec == null) {
            System.out.println("Unsupported codec for video file");
            throw new IllegalStateException();
        }
        ret = avcodec_open2(codec_ctx, codec, (PointerPointer)null);
        if (ret < 0) {
            System.out.println("Can not open codec");
            throw new IllegalStateException();
        }

        avutil.AVFrame frm = av_frame_alloc();

        // Allocate an AVFrame structure
        avutil.AVFrame pFrameRGB = av_frame_alloc();
        if (pFrameRGB == null) {
            System.exit(-1);
        }

        // Determine required buffer size and allocate buffer
        int numBytes = av_image_get_buffer_size(AV_PIX_FMT_RGB24, codec_ctx.width(),
                codec_ctx.height(), 1);
        BytePointer buffer = new BytePointer(av_malloc(numBytes));

        swscale.SwsContext sws_ctx = sws_getContext(
                codec_ctx.width(),
                codec_ctx.height(),
                codec_ctx.pix_fmt(),
                codec_ctx.width(),
                codec_ctx.height(),
                AV_PIX_FMT_RGB24,
                SWS_BILINEAR,
                null,
                null,
                (DoublePointer)null
        );

        if (sws_ctx == null) {
            System.out.println("Can not use sws");
            throw new IllegalStateException();
        }

        av_image_fill_arrays(pFrameRGB.data(), pFrameRGB.linesize(),
                buffer, AV_PIX_FMT_RGB24, codec_ctx.width(), codec_ctx.height(), 1);

        int frameCount = 0;
        long startTime = System.currentTimeMillis();

        i = 0;
        int ret1 = -1, ret2 = -1, fi = -1;
        while (av_read_frame(fmt_ctx, pkt) >= 0) {
            if (pkt.stream_index() == v_stream_idx) {
                ret1 = avcodec_send_packet(codec_ctx, pkt);
                ret2 = avcodec_receive_frame(codec_ctx, frm);

                //System.out.printf("ret1 %d ret2 %d\n", ret1, ret2);

                // avcodec_decode_video2(codec_ctx, frm, fi, pkt);
            }
            // if not check ret2, error occur [swscaler @ 0x1cb3c40] bad src image pointers
            // ret2 same as fi
            // if (fi && ++i <= 5) {
            if (ret2 >= 0) {
                sws_scale(
                        sws_ctx,
                        frm.data(),
                        frm.linesize(),
                        0,
                        codec_ctx.height(),
                        pFrameRGB.data(),
                        pFrameRGB.linesize()
                );

                //save_frame(pFrameRGB, codec_ctx.width(), codec_ctx.height(), i);
                // save_frame(frm, codec_ctx.width(), codec_ctx.height(), i);
            }
            av_packet_unref(pkt);

            ++frameCount;
        }

        float secs = (System.currentTimeMillis() - startTime) / 1000.0f;

        System.out.println("processing " + frameCount + " frames took " + (Math.round(secs * 1000) / 1000.0f) + " (" + (Math.round(frameCount / secs * 1000) / 1000.0f) + " fps)");

        av_frame_free(frm);

        avcodec_close(codec_ctx);
        avcodec_free_context(codec_ctx);

        avformat_close_input(fmt_ctx);

        System.out.println("Shutdown");
        System.exit(0);
    }
```

The only difference I spotted between your grabber and this other piece of code is you used a few deprecated calls in the grabber, where the other example uses new functions meant to replace the deprecated ones (`avcodec_decode_video2()` is deprecated and shoud be replaced with `avcodec_send_packet()` and `avcodec_receive_frame()`). But I don't believe this is the real issue.

**Comment by saudet** [comment]
Probably not an issue, no. Try to modify the rest, such as not decoding
audio and not converting images.

**Comment by b005t3r** [comment]
Ah, just noticed what caused the difference - ReadFewFrame also counts empty frames which are skipped in the grabber test. Processing times are the same for both, so there's no difference actually between those two (they are both about 3x slower than the command line app; about 2x slower when converting images is disabled).

I'm gonna create a new ticket for this tomorrow and describe what I found. Maybe it's simply the JNI's overhead, maybe something else, I won't be investigating this further anyway.

**Comment by saudet** [comment]
Do the times match by any chance if you omit ` -hwaccel videotoolbox` from the command line?

**Comment by b005t3r** [comment]
Nope, doesn't actually look like it's doing anything, so there's no difference in processing times.

**Comment by saudet** [comment]
I think I found the issue. FFmpeg doesn't seem to activate multithreading for decoding by default, but we should be able to use it by setting thread_count and thread_type like this:
https://stackoverflow.com/questions/43251612/ffmpeg-how-to-use-multithreading

**Comment by b005t3r** [comment]
You might be right. I modified the `ReadFewFrame.java` example and added this:

```
        avcodec.AVCodecContext codec_ctx = avcodec_alloc_context3(null);
        // use multithreading
        codec_ctx.thread_count(8);
        codec_ctx.thread_type(FF_THREAD_FRAME);
```

And it's basically on par with the command line version (about 2x faster than it used to be). Success :)

Would be great if you added this as a param to `FFmpegFrameGrabber`, but it looks like this might be the thing.

**Comment by saudet** [comment]
Could you check how the ffmpeg program sets those variables and we'll do the same?

**Comment by b005t3r** [comment]
It probably makes little sense to set it to an arbitrary number, as the optimal value depends on the number of cores, doesn't it? In my opinion, the best approach here would be to add a param to `FFmpegFrameGrabber` for settings the number of threads used.

On my laptop I've got 2 cores with hyper threading, so 4 independent threads in theory, so setting number of threads to 5 would be the optimum value (I just checked that, above 5 threads it doesn't get any faster), but on desktops it might be possible to get an extra boost with higher values.

As for the `thread_type`, it runs faster only with this setting `FF_THREAD_FRAME` (but I'd also suggest making this a separate param).

**Comment by saudet** [comment]
Sounds reasonable, but let's do by default what the ffmpeg program does!

**Comment by b005t3r** [comment]
OK :)

https://superuser.com/questions/155305/how-many-threads-does-ffmpeg-use-by-default

They say it depends :) Sometimes it's 1 thread per core, for lib264x it's 1.5 thread per core. I'd stick with the general rule of thumb for multithreading, which is:

```java
int cores = Runtime.getRuntime().availableProcessors(); // includes hyper threading cores
int threads = cores + 1;
```

:)

**Comment by saudet** [comment]
Yeah, no, let's not do that. Please try to find what the ffmpeg program
does, or let's leave it off by default.

**Comment by b005t3r** [comment]
OK, listen, try reading through my comments at least once, OK? I understand you're busy and all, but believe me, I'm busy, too and having to repeat the same thing over and over again is seriously frustrating.

### IT DEPENDS ON THE CODEC USED

There's no one value/formula, different number of threads is used for different codecs, as explained here:

https://superuser.com/questions/155305/how-many-threads-does-ffmpeg-use-by-default

This will also be different for different applications/systems/solutions, etc. I can hardly imagine someone wanting to run it at 12 threads by default, for instance. But I can also hardly imagine someone wanting to make it run 3x slower with only 1 thread for any logical reasons, so being able to actually set it up **consciously** is in my opinion more important than some arbitrary value no one knows about - add it as a constructor param maybe, this should make it obvious enough for everyone it's something they have to consider.

**Comment by saudet** [comment]
That post only mentions the libx264 codec, what do you want to do for all the other 50 ones? If you really want to figure out, go figure it out! I'm not going to make FFmpegFrameGrabber buggy or hard to use by default just because you do not want to have any custom code in your application! Am I being clear?

**Comment by b005t3r** [comment]
OMG :) That's exactly my point, it's impossible to tell how many is just right :) Just make it a param in the constructor so anyone can choose what they see fit.

**Comment by saudet** [comment]
A new constructor parameter for something this obscure is only going to make it harder to use! I'll just get more people complaining like you, and you won't be there to help, so no! I'm the one that has to deal with people like you, and I don't want more.

Now, it looks like if we set thread_count to 0, leaving thread_type uninitialized, it kicks into auto mode, so that's probably what the ffmpeg program does. Unless you actually want to help and figure out something better, I think that's what I will do.

**Comment by b005t3r** [comment]
I said what I want - I want to be able to set it by myself. Whatever you set it by default - I really don't care that much.

**Comment by saudet** [comment]
We can already set it with the "threads" option like this:
```
        grabber.setVideoOption("threads", "8");
        grabber.setAudioOption("threads", "8");
```
You would have already known that if you had taken a look at the relevant documentation:
https://ffmpeg.org/ffmpeg-codecs.html

**Comment by b005t3r** [comment]
OMG, it's been a whole week of arguing and only NOW you give me this answer? What's wrong with you? Why have this whole argument, look for APIs, ask me to check things, when you could have said that in the first place? I gave you the code I was using fro my test, why did you just say: "Hey, simply add a `grabber.setVideoOption("threads", "0");` thing and it will go much faster`. Why? Why go through how it's implemented in ffmpeg tool, why not set it by default and not use it in any of your examples until now? I don't get you, man. Seriously, thanks for the tip, but, for the next time, maybe don't make people go through a week of testing stuff you know is unnecessary and simply give them the answer you already know OR don't be like "_you would have already known that if you had taken a look at the relevant documentation_", because you also hadn't done that until yesterday and only after looking elsewhere (and making someone else look elsewhere) for a week!

**Comment by saudet** [comment]
Reread the thread, I didn't know the issue was about threads at first. When I did figure out, because you were not motivated to figure out on your own, your only concern was about the default behavior or about adding a parameter to the constructor. If you don't care about the default behavior, then make no mention of it. Stop complaining so much, and start helping others help you!

**Comment by mifritscher** [comment]
Sorry to drive in - but...

Does the line
```
grabber.setVideoOption("threads", "1"); // more precise without threads
```
in the test really overwrite the 
```
video_c.thread_count(0);
```
? If yes, perhaps it would be nice to have a comment at the options-for loop "this can overwrite all setting done in video/audio_c". And shifting the ```video_c.thread_count(0);``` above the option loop to further clear the priorities. Because as of now, one could think that this option can't be overwritten at the first glance.

Also, to be honest, I've a bit of fear changing the default behavior regarding threads. Perhaps a new function "set[Audio|Video]Threads()" with a javadoc mentioning 0="select number of threads automatically" is a better way. This will let stumble people over it when they look over the API for optimizing the throughput, but does not break existing code (You had even to adapt your test case...)
The ffmpeg default (using 1...1.5 threads per core on most codecs) can make sense in command line tools, but for my taste not in libraries which are often used in more complicated cases. For example, I use javacv for transcoding up to 8 streams in parallel on a 4 core machine. I don't think that it would be particularly useful having 32...48 threads on it ;-)

**Comment by b005t3r** [comment]
@mifritscher, had exactly the same doubts, but yes, it does override it, at least my tests showed it does. But yeah, the code, as it is now, suggests it might not.

**Comment by saudet** [comment]
@mifritscher Feel free to send a PR that includes more documentation! As for changing the default, libx264 and others were already using threads, it's just now more consistent across codecs. To disable threading for libx264, we already had to set "threads" to "1".

**Comment by saudet** [comment]
Anyway, if there's any other issues related to threads, open a new issue or pull request as appropriate. Thanks for pointing out this issue and for persisting!
